patches/api/0423-Paper-Item-API.patch

```diff
@@ -0,0 +1,653 @@
++                T val = this.getValue(key);
++                if (val == null) {
++                    return value;
++                }
++
++                return val;
++            }
++
++            @Override
++            public boolean hasProperty(PropertyKey<?> propertyKey) {
++                return ImmutablePropertyBuilder.this.map.containsKey(propertyKey);
++            }
++        };
++    }
++
++
++    private record Entry<T>(PropertyKey<? extends Value<T>> key, T value) {
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/property/PropertyHolder.java b/src/main/java/io/papermc/paper/property/PropertyHolder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3d281c081d3a997455d19e9dd5e86d1806c830b6
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/property/PropertyHolder.java
+@@ -0,0 +1,33 @@
++package io.papermc.paper.property;
++
++import io.papermc.paper.property.value.MapValue;
++import io.papermc.paper.property.value.Value;
++import org.bukkit.enchantments.Enchantment;
++
++import java.util.Map;
++
++// Represents a property reader during runtime
++public interface PropertyHolder extends PropertyReader {
++
++    interface Mutable extends PropertyHolder, Cloneable {
++
++        void set(Value<?> value);
++
++        <T> void unset(PropertyKey<? extends Value<T>> key);
++
++        <T> void set(PropertyKey<? extends Value<T>> key, T value);
++
++        <V, T extends Value<V>> T create(PropertyKey<T> key);
++
++        <V, T extends Value<V>> T getOrCreate(PropertyKey<T> key);
++    }
++
++    interface Immutable extends PropertyHolder {
++
++        public static ImmutablePropertyBuilder builder() {
++            return new ImmutablePropertyBuilder();
++        }
++
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/property/PropertyKey.java b/src/main/java/io/papermc/paper/property/PropertyKey.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d18841d965c239973068b33e5b187929df0648ee
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/property/PropertyKey.java
+@@ -0,0 +1,72 @@
++package io.papermc.paper.property;
++
++import com.google.common.collect.Multimap;
++import io.leangen.geantyref.TypeFactory;
++import io.leangen.geantyref.TypeToken;
++import io.papermc.paper.property.value.ListValue;
++import io.papermc.paper.property.value.MapValue;
++import io.papermc.paper.property.value.MultiMapValue;
++import io.papermc.paper.property.value.SetValue;
++import io.papermc.paper.property.value.Value;
++import net.kyori.adventure.text.Component;
++import org.bukkit.attribute.Attributable;
++import org.bukkit.attribute.AttributeModifier;
++
++import java.lang.reflect.Type;
++import java.util.List;
++import java.util.Map;
++import java.util.Set;
++
++public interface PropertyKey<T extends Value<?>> {
++
++    static <V> PropertyKey<SetValue<V>> ofSet(String key, TypeToken<V> valueToken) {
```

Comment:
The issue I see here currently, is that any dev is able to define arbitrary nbt entries.
They could probably create an invalid map itemstack with it.
Instead of 

```java
 PropertyKey<Value<Integer>> MAP_ID = PropertyKey.of("map", Integer.class);
```
they could for example use that.

```java
 PropertyKey<Value<String>> MAP_ID_CODE = PropertyKey.of("map", String.class);
 ```

The game / client expects the `map` to be an Integer but it gets a String instead.
As with an interface we cannot ensure package-private static methods, it should either be moved to another class.

A Bukkitvalues equivalent would be neat as well.

Reply:
I would just argue that creation of such keys should be blocked somehow? only the api itself is allowed to create keys

Reply:
@MiniDigger 

I discussed it a bit in the paper contributing thread but I disagree with this. In general I think that in general devs being able to add their own keys does no harm, but it won’t be officially supported. You’ll need to register your own adapter for that key, which will require NMS. So it’s technically blocked normally but is still doable with access to sources.

Reply:
Decided against in the end